example/js/api/diplomat-runtime.mjs

```diff
@@ -182,15 +182,102 @@ export class DiplomatBuf {
 	}
 
 	constructor(ptr, size, free) {
-	this.ptr = ptr;
-	this.size = size;
-	// Generated code calls one of methods these for each allocation, to either
-	// free directly after the FFI call, to leak (to create a &'static), or to
-	// register the buffer with the garbage collector (to create a &'a).
-	this.free = free;
-	this.leak = () => { };
-	this.garbageCollect = () => DiplomatBufferFinalizer.register(this, this.free);
+		this.ptr = ptr;
+		this.size = size;
+		// Generated code calls one of methods these for each allocation, to either
+		// free directly after the FFI call, to leak (to create a &'static), or to
+		// register the buffer with the garbage collector (to create a &'a).
+		this.free = free;
+		this.leak = () => { };
+		this.garbageCollect = () => DiplomatBufferFinalizer.register(this, this.free);
 	}
 }
 
+export class DiplomatWriteBuf {
+	leak;
+
+	#wasm;
+	#buffer;
+
+	constructor(wasm) {
+		this.#wasm = wasm;
+		this.#buffer = this.#wasm.diplomat_buffer_write_create(0);
+
+		this.leak = () => { };
+	}
+	
+	free() {
+		this.#wasm.diplomat_buffer_write_destroy(this.#buffer);
+	}
+
+	garbageCollect() {
+		DiplomatBufferFinalizer.register(this, this.free);
+	}
+
+	readString8() {
+		return readString8(this.#wasm, this.ptr, this.size);
+	}
+
+	get buffer() {
+		return this.#buffer;
+	}
+
+	get ptr() {
+		return this.#wasm.diplomat_buffer_write_get_bytes(this.#buffer);
+	}
+
+	get size() {
+		return this.#wasm.diplomat_buffer_write_len(this.#buffer);
+	}
+}
+
+export class DiplomatReceiveBuf {
+	#wasm;
+
+	#size;
+	#align;
+
+	#hasResult;
+
+	#buffer;
+
+	constructor(wasm, size, align, hasResult) {
+		this.#wasm = wasm;
+
+		this.#size = size;
+		this.#align = align;
+
+		this.#hasResult = hasResult;
+
+		this.#buffer = this.#wasm.diplomat_alloc(this.#size, this.#align);
+
+		this.leak = () => { };
+	}
+	
+	free() {
+		this.#wasm.diplomat_free(this.#buffer, this.#size, this.#align);
+	}
+
+	garbageCollect() {
```

Comment:
and if this is only for structs, this method should never be necessary I think? The backend always immediately destructures structs so there is no need to GC the aggregate.

Reply:
looks like even though it's also used for slices, this method is still not necessary since we currently appear to copy slices over using `sliceFromPtr`.

thought 1: the `sliceFromPtr` stuff could be made a method on this type to clean up callsites

thought 2: in JS when we return slices _ought_ we be returning slices into WASM memory instead? We could, if we started returning ArrayBuffers, however we would have to attack borrow edges, whereas here we do not.

Reply:
I've removed `garbageCollect`, and moved the slice functions to `DiplomatReceiveBuf`. As for the last point, I'll make it a new issue.

Reply:
#602